file_cache.py: idempotent remove without races

haampie opened the pull request:
Closes #31475

There's a bunch of issues with `.remove()`:

- Deleting the lock file after releasing the lock implies race conditions (another process may have acquired it):
   ```
   ==> Error: Attempted to close non-existing lock path: /home/harmen/.spack/cache/indices/.62a4a153a8_c373ec5d29.json.lock
   ```
- When multiple processes call `remove` on the same file, all but one error because the file is not found, which happens during parallel spack install on a fresh spack (binary cache has some delayed call to remove() which almost always runs into this).
- ~Releasing and removing the lock file runs into issues when a single processes holds read/write locks.~ (Maybe this isn't so bad, cause on Windows you can't delete a file if you have an open file descriptor? On Linux the file is unlinked but the inode remains... so not a very portable use case I guess)

The simplest solution:
- make .remove() idempotent so multiple deletes don't error
- don't delete locks so there are no races

alalazo commented on lib/spack/spack/util/file_cache.py:
```diff
@@ -173,7 +173,10 @@ def remove(self, key):
-            os.unlink(self.cache_path(key))
+            try:
+                os.unlink(self.cache_path(key))
+            except FileNotFoundError:
```
If only we supported Python >= 3.6 :sweat:

haampie committed:
file_cache.py: idempotent remove without races

There's a race condition in `remove()` as the lockfile is removed after
releasing the lock, which is a problem when another process acquires a
write lock during deletion.

Also simplify life a bit in multiprocessing when a file is possibly
removed multiple times, which currently is an error on the second
deletion, so the proposed fix is to make remove(...) idempotent and not
error when deleting non-existing cache entries.

haampie commented:
With this PR, the following would save `hello` with the contents `second` without erroring (since the file is persisted only at the end of the write transaction).

```python
#!/usr/bin/env spack-python

from spack.util.file_cache import FileCache

cache = FileCache(".")

with cache.write_transaction("hello") as (old, new):
    new.write("first\n")

with cache.write_transaction("hello") as (old, new):
    new.write("second\n")
    cache.remove("hello")
    cache.remove("hello")
```

On `develop`, it would error out because `cache.remove` tries to delete a lock that's in use.

haampie committed:
Don't tests for existence of lockfile, cause windows/linux behavior is different

haampie committed:
use errno

alalazo commented on lib/spack/spack/test/util/file_cache.py:
```diff
@@ -94,3 +97,9 @@ def test_cache_write_readonly_cache_fails(file_cache):
         file_cache.write_transaction(filename)
+
+
+def test_delete_is_idempotent(file_cache):
```
```suggestion
@pytest.mark.regression('31475')
def test_delete_is_idempotent(file_cache):
```

alalazo reviewed (changes requested):
Just a minor comment, but otherwise LGTM

haampie committed:
Update lib/spack/spack/test/util/file_cache.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>
